File: homing-studio-base/src/main/resources/homing/js/hue/captains/singapura/js/homing/studio/base/ui/layout/FocusScopeModule.js
// =============================================================================
// FocusScopeModule — RFC 0048 give-up-focus wrapper.
//
// A container that owns a "you are inside me now" mode wraps its content in a
// FocusScope and is told, uniformly, when the content wants to hand focus back.
// It normalises the two give-up signals on `hostEl` (bubble phase) into one
// onGiveUp():
//
//   1. an un-consumed Escape  — a keydown that bubbled to hostEl because no
//      widget inside called stopPropagation; or
//   2. a release event        — a bubbling CustomEvent('homing-focus',
//      { detail: { intent: 'release' } }).
//
// The dual signal is trap-free without stealing a key: a widget that uses Escape
// keeps it (stopPropagation); a widget that swallows Escape entirely binds its
// own shortcut to dispatch the release event. Same class in every container, so
// give-up-focus is uniform — a portable DOM convention, not an injected API.
//
// Builds no DOM and applies no styling — it only attaches two listeners.
//
// API:  new FocusScope(hostEl, onGiveUp).attach();  … .dispose();
// =============================================================================

class FocusScope {

    constructor(hostEl, onGiveUp) {
        if (!hostEl) throw new Error("[FocusScope] hostEl is required");
        if (typeof onGiveUp !== "function") throw new Error("[FocusScope] onGiveUp must be a function");
        this._host     = hostEl;
        this._onGiveUp = onGiveUp;
        this._attached = false;
        var self = this;
        // Un-consumed Escape — anything inside that uses Escape would have called
        // stopPropagation, so its arrival here IS the "not consumed" signal.
        this._onKey = function (e) {
            if (e.key === "Escape" && !e.ctrlKey && !e.metaKey && !e.altKey) self._give(e);
        };
        // Voluntary release — a widget dispatches this on its own trigger (a Done
        // button, task completion, or a custom shortcut when it owns Escape).
        this._onRelease = function (e) {
            var d = e && e.detail;
            if (d && d.intent === "release") self._give(e);
        };
    }

    attach() {
        if (this._attached) return this;
        this._host.addEventListener("keydown", this._onKey);            // bubble phase
        this._host.addEventListener("homing-focus", this._onRelease);   // bubble phase
        this._attached = true;
        return this;
    }

    dispose() {
        if (!this._attached) return this;
        this._host.removeEventListener("keydown", this._onKey);
        this._host.removeEventListener("homing-focus", this._onRelease);
        this._attached = false;
        return this;
    }

    _give(e) {
        // Consume it so an outer FocusScope (nested containers) doesn't also fire.
        if (e) {
            if (typeof e.stopPropagation === "function") e.stopPropagation();
            if (typeof e.preventDefault === "function")  e.preventDefault();
        }
        try { this._onGiveUp(); }
        catch (err) { console.error("[FocusScope] onGiveUp threw:", err); }
    }
}

File: homing-studio-base/src/main/java/hue/captains/singapura/js/homing/studio/base/ui/layout/FocusScopeModule.java
package hue.captains.singapura.js.homing.studio.base.ui.layout;

import hue.captains.singapura.js.homing.core.DomModule;
import hue.captains.singapura.js.homing.core.Exportable;
import hue.captains.singapura.js.homing.core.ExportsOf;
import hue.captains.singapura.js.homing.core.ImportsFor;

import java.util.List;

/**
 * RFC 0048 — {@code FocusScope}: a small, reusable <b>give-up-focus wrapper</b>.
 * A container that owns a "you are inside me now" mode (a
 * {@link MultiTabPaneModule MultiTabPane} pane entered deep, a modal, …) wraps
 * its content in a {@code FocusScope} and is told, uniformly, when the content
 * wants to hand focus back.
 *
 * <p>The whole point is that a widget must work <b>anywhere</b>, so the wrapper
 * imposes <b>nothing special</b> on it — it rides only the two mechanisms every
 * DOM element already has. {@code new FocusScope(hostEl, onGiveUp)} listens on
 * {@code hostEl} (bubble phase) and calls {@code onGiveUp()} on <b>either</b>:</p>
 *
 * <ol>
 *   <li>an <b>un-consumed {@code Escape}</b> — a {@code keydown} that bubbled all
 *       the way up because no widget inside called {@code stopPropagation}; or</li>
 *   <li>a <b>release event</b> — a bubbling
 *       {@code CustomEvent('homing-focus', { detail: { intent: 'release' } })}.</li>
 * </ol>
 *
 * <p>The dual signal is what makes it trap-free without the container ever
 * stealing a key: a widget that uses {@code Escape} for its own function keeps
 * it (standard {@code stopPropagation}); a widget that <em>swallows Escape
 * entirely</em> (a terminal, an Excel-like grid) instead binds its own shortcut
 * to {@code dispatchEvent} the release event. Both routes reach the same wrapper.
 * Because every container uses this one class, give-up-focus behaves identically
 * everywhere — a portable DOM convention, not an injected API.</p>
 *
 * <p>Dependency-free; builds no DOM and applies no styling (it only attaches two
 * listeners), so it carries no CSS/DOM-ownership obligations of its own.</p>
 */
public record FocusScopeModule() implements DomModule<FocusScopeModule> {

    /** The single export — the {@code FocusScope} JS class. */
    public record FocusScope() implements Exportable._Constant<FocusScopeModule> {}

    public static final FocusScopeModule INSTANCE = new FocusScopeModule();

    @Override
    public ImportsFor<FocusScopeModule> imports() {
        return ImportsFor.noImports();
    }

    @Override
    public ExportsOf<FocusScopeModule> exports() {
        return new ExportsOf<>(INSTANCE, List.of(new FocusScope()));
    }
}
